meson.build

```diff
@@ -0,0 +1,117 @@
+project('profanity',
+        'c',
+        license: 'GPL3+',
+        default_options:  ['c_std=gnu99'],
+        version: '0.12.0',
+        meson_version: '>= 0.54.0')
+
+
+cc = meson.get_compiler('c')
+conf = configuration_data()
+
+
+# general dependencies
+glib_dep = dependency('glib-2.0', version: '>= 2.62.0')
+gio_dep = dependency('gio-2.0')
+curl_dep = dependency('libcurl', version: '>= 7.62.0')
+sqlite_dep = dependency('sqlite3', version: '>= 3.22.0')
+thread_dep = dependency('threads') # pthread
+thread_dep = dependency('readline')
+m_dep = cc.find_library('m', required : true)
+
+
+# this will detect ncursesw etc (-I / -l flags)
+curses_dep = dependency('curses', required: false)
+# get ncurses definitions
+check_headers = [
+  ['ncursesw/menu.h', 'HAVE_NCURSESW_MENU_H'],
+  ['ncurses/menu.h', 'HAVE_NCURSES_MENU_H'],
+  ['menu.h', 'HAVE_MENU_H'],
+  ['ncursesw/curses.h', 'HAVE_NCURSESW_CURSES_H'],
+  ['ncursesw.h', 'HAVE_NCURSESW_H'],
+  ['ncurses/curses.h', 'HAVE_NCURSES_CURSES_H'],
+  ['ncurses.h', 'HAVE_NCURSES_H'],
+  ['curses.h', 'HAVE_CURSES_H'],
+]
+
+foreach h : check_headers
+  if cc.has_header(h.get(0))
+    conf.set(h.get(1), 1)
+  endif
+endforeach
+
+
+libstrophe_dep = dependency('libstrophe', version: '>= 0.11.0')
+conf.set('HAVE_LIBSTROPHE', 1)
+
+
+# optional dependencies
+notify_dep = dependency('libnotify', required : get_option('notifications'))
+if notify_dep.found()
+  conf.set_quoted('HAVE_LIBNOTIFY', '1')
+endif
+# TODO: check for otr3 and otr4
+otr_dep = dependency('libotr', version: '>= 4.0', required : get_option('otr'))
+if otr_dep.found()
+  conf.set_quoted('BUILD_OTR', '1')
+endif
+pgp_dep = dependency('gpgme', required : get_option('pgp'))
+if pgp_dep.found()
+  conf.set_quoted('BUILD_PGP', '1')
```

Comment:
"notifications" is a feature object. When configured with `-Dnotifications=disabled` it logs:

```
Dependency libnotify skipped: feature notifications disabled
[...]
    notifications      : disabled
```
and builds without support.

gpg is a boolean option, which IMHO it shouldn't be. When configured with `-Dpgp=false` it logs:

```
Run-time dependency gpgme found: YES 1.16.0
[...]
    pgp                : false
```

which is contradictory and misleading, because it was available and detected, just not *required*. So it tries to build with PGP support.

BUILD_PGP is the Makefile.am define, not the C define. This should set HAVE_LIBGPGME.

Later, in `core_sources_c`, check if pgp_dep.found() and add 'src/pgp/gpg.c' as an additional source.

Reply:
So basically in our case all the options in meson_options.txt should be `type: feature` right?

Reply:
Not entirely.

Feature options are tristate options, booleans are dual-state.
- When used in `if get_option('foobar')` then booleans are on/off switches.
- When used as `required: get_option('foobar')`, then booleans act like "enabled"/"auto" switches. In this case there is no off (disabled) switch.

If you aren't looking up a dependency, just enabling parts of the build, then booleans are correct. For example, c-plugins is either on or off, it has no dependencies so "auto" makes no sense.

You can also do booleans like this:

```meson
if get_option('pgp')
    pgp_dep = dependency('gpgme')
else
    # dummy always-not-found dependency
    pgp_dep = dependency('', required: false)
endif
```

This would be either enabled or disabled, with no "auto" option. But auto is nice.